src/marqo/tensor_search/telemetry.py

```diff
@@ -0,0 +1,214 @@
+        
+        Return time is in Ms.
+        """
+        if self.start_time is None:
+            logger.warn(f"'.stop()' called on unstarted timer. '.start()' must be called before '.stop()'.")
+            raise TimerError()
+        else:
+            elapsed_time = time.perf_counter() - self.start_time
+            self.start_time = None
+            return 1000 * elapsed_time
+
+
+class RequestMetric:
+    @classmethod
+    def reduce_from_list(cls, metrics: List["RequestMetric"]) -> "RequestMetric":
+        assert len(metrics) > 0, "Cannot create RequestMetric from []"
+        m = metrics.pop(0)
+        for mm in metrics:
+            for k, count in mm.counter.items():
+                m.increment_counter(k, count)
+
+            for k, timer in mm.timers.items():
+                m.timers[k] = timer
+
+            for k, time in mm.times.items():
+                m.add_time(k, time)
+        return m
+
+    def __init__(self):
+        self.counter: Dict[str, int] = defaultdict(int)
+
+        # Note: We default to float, but on multiple times for the same key, it gets converted to a List.
+        self.times: Dict[str, Union[float, List[float]]] = defaultdict(float)
+        self.timers: Dict[str, Timer] = defaultdict(Timer)
+
+    def increment_counter(self, k: str):
+        self.counter[k]+=1
+
+    @contextmanager
+    def time(self, k: str, callback: Optional[Callable[[float], None]] = None):
+        self.start(k)
+        try:
+            yield
+        finally:
+            elapsed_time = self._stop(k)
+            if callback is not None:
+                callback(elapsed_time)
+            self.add_time(k, elapsed_time)
+
+    def start(self, k: str):
+        """Start a new timer for the given key"""
+        self.timers[k].start()
+
+    def _stop(self, k: str) -> float:
+        return self.timers[k].stop()
+
+    def add_time(self, k: str, v: float):
+        if self.times.get(k, None) is None:
+            self.times[k] = v
+        elif isinstance(self.times.get(k), list):
+            self.times[k] = self.times[k] + [v]
+        else:
+            self.times[k] = [self.times[k], v]
+
+    def stop(self, k: str) -> float:
+        """Stop the timer for the given key, and report the elapsed time"""
+        try:
+            elapsed_time = self._stop(k)
+            self.add_time(k, elapsed_time)
+            return elapsed_time
+        except TimerError:
+            logger.warn(f"timer {k} stopped incorrectly. Time not recorded.")
+
+    def increment_counter(self, k: str, v: int = 1):
+        self.counter[k]+=v
+
+    def json(self):
+        return {
+            "counter": dict(self.counter),
+            "timesMs": dict(self.times)
+        }
+
+
+class RequestMetrics():
+    current_request: ContextVar[Request] = ContextVar('current_request')
+    
+    METRIC_STORES: Dict[str, RequestMetric] = {}
+
+    @classmethod
+    def _set_request(cls, r: Request):
+        cls.current_request.set(r)
```

Comment:
I wonder if `cls.current_request` could pose a concurrency risk.

Reply:
This did concern us. Because starlette (the async ASGI framework behind fastapi) doesn't run a Request across threads, it should not be touched across threads. Testing didn't seem to uncover anything critical (even though this is a null result, in a sense). Even with that, we set it within middleware and for specific thread operations (not override). I shall add a comment that this is something that should be used liberally.